provider/equinix/environ.go

```diff
@@ -0,0 +1,779 @@
+	opt := &packngo.ListOptions{
+		Includes: []string{"available_in_metros"},
+	}
+	plans, _, err := e.equinixClient.Plans.List(opt)
+	if err != nil {
+		return nil, errors.Annotate(err, "retrieving supported instance types")
+	}
+
+	var instTypes []instances.InstanceType
+nextPlan:
+	for _, plan := range plans {
+		if !validPlan(plan, e.cloud.Region) {
+			logger.Tracef("Plan %s not valid in facility %s", plan.Name, e.cloud.Region)
+			continue
+		}
+
+		var instArch string
+		switch {
+		case strings.HasSuffix(plan.Name, ".x86"):
+			instArch = arch.AMD64
+		case strings.HasSuffix(plan.Name, ".arm"):
+			instArch = arch.ARM64
+		default:
+			continue nextPlan
+
+		}
+
+		mem, err := parseMemValue(plan.Specs.Memory.Total)
+		if err != nil {
+			continue
+		}
+
+		instTypes = append(instTypes,
+			instances.InstanceType{
+				Id:       plan.ID,
+				Name:     plan.Name,
+				CpuCores: uint64(plan.Specs.Cpus[0].Count),
+				Mem:      mem,
+				Arches:   []string{instArch},
+				// Scale per hour costs so they can be represented as an integer for sorting purposes.
+				Cost: uint64(plan.Pricing.Hour * 1000.0),
+				// The Equinix Metal API returns all plan as legacy today. There is an issue open internally to figure out why.
+				// In the meantime let's comment this https://github.com/juju/juju/pull/12983#discussion_r635324484
+				// Deprecated: plan.Legacy,
+			})
+	}
+
+	return instTypes, nil
+}
+
+func validPlan(plan packngo.Plan, region string) bool {
+	// some plans may not be servers
+	if plan.Pricing == nil ||
+		plan.Specs == nil ||
+		plan.Specs.Memory == nil ||
+		len(plan.Specs.Cpus) == 0 || plan.Specs.Cpus[0].Count == 0 {
+		return false
+	}
+	for _, a := range plan.AvailableInMetros {
+		// some plans are not available in-region
+		if a.Code != region {
+			continue
+		}
+		return true
+	}
+	return false
+}
+
+func parseMemValue(v string) (uint64, error) {
+	var scaler = uint64(1)
+	if strings.HasSuffix(v, "GB") {
+		scaler = 1024
+		v = strings.TrimSuffix(v, "GB")
+	}
+
+	val, err := strconv.ParseUint(v, 10, 64)
+	return val * scaler, err
+}
+
+func (e *environ) findInstanceSpec(controller bool, allImages []*imagemetadata.ImageMetadata, instanceTypes []instances.InstanceType, ic *instances.InstanceConstraint) (*instances.InstanceSpec, error) {
+	oss, _, err := e.equinixClient.OperatingSystems.List()
+	if err != nil {
+		return nil, err
+	}
+	var suitableImages []*imagemetadata.ImageMetadata
+
+	for _, it := range instanceTypes {
+		for _, os := range oss {
+			if !isDistroSupported(os, ic) {
+				continue
+			}
+
+			for _, p := range os.ProvisionableOn {
+				if p == it.Name {
+					image := &imagemetadata.ImageMetadata{
+						Id:   os.Slug,
+						Arch: getArchitectureFromPlan(p),
+					}
+					suitableImages = append(suitableImages, image)
+				}
+			}
+		}
+	}
+
+	images := instances.ImageMetadataToImages(suitableImages)
+	spec, _ := instances.FindInstanceSpec(images, ic, instanceTypes)
+	return spec, err
```

Comment:
If we don't find the instance spec we should return the error. Now I don't know why we get an error here some of the time, but we do.